File: nandefact-api/src/domain/factura/MontoIVA.ts
import type { TasaIVA } from '../shared/types.js';
import { MontoInvalidoError } from '../errors/MontoInvalidoError.js';

/**
 * Value Object — Cálculo de IVA sobre un monto en PYG.
 * Los precios en Paraguay INCLUYEN IVA. Calcula base gravada y monto IVA.
 * Todos los montos son enteros (PYG sin decimales).
 */
export class MontoIVA {
  readonly montoTotal: number;
  readonly tasaIVA: TasaIVA;
  readonly baseGravada: number;
  readonly montoIVACalculado: number;

  constructor(montoTotal: number, tasaIVA: TasaIVA) {
    if (!Number.isInteger(montoTotal)) {
      throw new MontoInvalidoError('el monto debe ser entero (PYG sin decimales)');
    }

    if (montoTotal < 0) {
      throw new MontoInvalidoError('el monto no puede ser negativo');
    }

    this.montoTotal = montoTotal;
    this.tasaIVA = tasaIVA;

    if (tasaIVA === 10) {
      this.baseGravada = Math.round(montoTotal / 1.1);
      this.montoIVACalculado = montoTotal - this.baseGravada;
    } else if (tasaIVA === 5) {
      this.baseGravada = Math.round(montoTotal / 1.05);
      this.montoIVACalculado = montoTotal - this.baseGravada;
    } else {
      // Exenta
      this.baseGravada = 0;
      this.montoIVACalculado = 0;
    }
  }

  equals(other: MontoIVA): boolean {
    return this.montoTotal === other.montoTotal && this.tasaIVA === other.tasaIVA;
  }
}

File: nandefact-app/shared/src/commonMain/kotlin/py/gov/nandefact/shared/domain/MontoIVA.kt
package py.gov.nandefact.shared.domain

/**
 * Calcula base gravada y monto IVA para PYG (sin decimales).
 * Los precios en Paraguay INCLUYEN IVA.
 */
object MontoIVA {

    data class Resultado(
        val baseGravada: Long,
        val montoIva: Long
    )

    /**
     * @param precioConIva Precio total incluyendo IVA
     * @param tasaIva 10, 5, o 0 (exenta)
     * @param proporcion 100 (normal), 85 (turismo), 30 (inmuebles)
     */
    fun calcular(
        precioConIva: Long,
        tasaIva: Int,
        proporcion: Int = 100
    ): Resultado {
        if (tasaIva == 0) {
            return Resultado(baseGravada = 0, montoIva = 0)
        }

        // Parte gravada según proporción
        val montoGravado = precioConIva * proporcion / 100
        val montoExento = precioConIva - montoGravado

        // Cálculo IVA sobre la parte gravada
        val baseGravada = montoGravado * 100 / (100 + tasaIva)
        val montoIva = montoGravado - baseGravada

        return Resultado(baseGravada = baseGravada, montoIva = montoIva)
    }
}
